# Block 1

In [7]:
# i will import the necessary libraries
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

# i will import the data
data = sns.load_dataset('geyser')

# Block 2

In [8]:
# i will create my own k-fold cross validation code
training_score = []
testing_score = []

value = len(data)*.10
# put eveyrthing into the same for loop (below)
# append r2 scores to a list and average them
for i in range(10):
  all_rows = np.arange(len(data)).tolist()
  testing_index = np.arange(int(value*i), int(value*(i+1)))
  testing_data = data.iloc[testing_index,:]
  for j in testing_index:
    all_rows.remove(j)
  training_index =  all_rows
  training_data = data.iloc[training_index,:]
  
  Xtest =  testing_data[['duration']]
  ytest = testing_data['waiting']

  Xtrain =  training_data[['duration']]
  ytrain = training_data['waiting']

  model = LinearRegression()
  model.fit(Xtrain, ytrain)

  testing_score.append(model.score(Xtest, ytest))
  training_score.append(model.score(Xtrain, ytrain))

# Block 3

In [9]:
# i will print the average and standard deviation of the r^2 values for the different folds
# testing scores
print('The mean of the testing R^2 values is: ', np.mean(testing_score))
print('The std of the testing R^2 values is: ', np.std(testing_score))
# training scores
print('The mean of the training R^2 values is: ', np.mean(training_score))
print('The std of the training R^2 values is: ', np.std(training_score))

The mean of the testing R^2 values is:  0.7974898504962091
The std of the testing R^2 values is:  0.06759008267092219
The mean of the training R^2 values is:  0.8115522476177027
The std of the training R^2 values is:  0.005330716057146054


# Block 4

Using my code, I have found that the mean of all my testing values is approximately .80 and the standard deviation is approximately .07. The mean of the training r squared value is approximately .81 and the standard deviation is approximately .01. This means that the r squared values for the training model is more reliable than the testing value, which is expected because we used the training data to train the model; however, the standard deviation for the testing data shows that the cross-validation model is still relatively reliable for the testing data. Therefore, I can conclude that my model can consistently use the duration of an eruption to predict the waiting time for the next eruption.